tt-train/tests/core/n300_utils_test.cpp

```diff
@@ -0,0 +1,171 @@
+    auto reduced_tensor = xtensor + xtensor;
+
+    EXPECT_TRUE(xt::allclose(reduced_tensor, xtensors_back[0]));
+    EXPECT_TRUE(xt::allclose(reduced_tensor, xtensors_back[1]));
+}
+
+TEST_F(N300UtilsTest, TestXTensorShardAxis2AddScalar) {
+    auto* device = &ttml::autograd::ctx().get_device();
+    auto mesh_shape = device->shape();
+    float scalar = 10.F;
+    xt::xarray<float> test_data = xt::arange(8);
+    xt::xarray<float> xtensor = test_data.reshape({1, 1, 2, 4});
+
+    ttml::core::XTensorToMeshVariant<float> shard_composer = ttml::core::ShardXTensorToMesh<float>(mesh_shape, 2);
+    auto tensor = ttml::core::from_xtensor(xtensor, device, shard_composer);
+    auto out_tensor = ttnn::add(tensor, scalar);
+    ttml::core::MeshToXTensorVariant<float> identity_composer = ttml::core::VectorMeshToXTensor<float>(mesh_shape);
+    auto xtensors_back = ttml::core::to_xtensor(out_tensor, identity_composer);
+
+    xt::xarray<float> chunk0 = xt::view(xtensor, xt::all(), xt::all(), xt::range(0, 1), xt::all());
+    xt::xarray<float> chunk1 = xt::view(xtensor, xt::all(), xt::all(), xt::range(1, 2), xt::all());
+
+    EXPECT_TRUE(xt::allclose(chunk0 + scalar, xtensors_back[0]));
+    EXPECT_TRUE(xt::allclose(chunk1 + scalar, xtensors_back[1]));
+}
+
+TEST_F(N300UtilsTest, TestXTensorShardAxis3Matmul) {
+    xt::random::seed(42);
+    auto* device = &ttml::autograd::ctx().get_device();
+    auto mesh_shape = device->shape();
+
+    xt::xarray<float> xtensor_a = xt::random::rand({128 * 64}, -0.005, 0.005).reshape({1, 1, 128, 64});
+    xt::xarray<float> xtensor_b = xt::random::rand({256 * 64}, -0.005, 0.005).reshape({1, 1, 64, 256});
+
+    ttml::core::XTensorToMeshVariant<float> replicate_composer2 = ttml::core::ShardXTensorToMesh<float>(mesh_shape, 2);
+    ttml::core::XTensorToMeshVariant<float> replicate_composer3 = ttml::core::ShardXTensorToMesh<float>(mesh_shape, 3);
+    auto tensor_a = ttml::core::from_xtensor(xtensor_a, device, replicate_composer3);
+    auto tensor_b = ttml::core::from_xtensor(xtensor_b, device, replicate_composer3);
+
+    auto gathered_ta =
+        ttnn::all_gather(tensor_a, 3 /*, {0, 4}, 1 ,std::nullopt, std::nullopt, std::nullopt, std::nullopt*/);
+    fmt::print("gathered_ta shape: {}\n", gathered_ta.get_shape().logical_shape());
+    auto mul_tensor = ttnn::matmul(
+        gathered_ta,
+        tensor_b,
+        false,
+        false,
+        /* memory_config */ std::nullopt,
+        /* dtype */ std::nullopt,
+        /* program_config */ std::nullopt,
+        /* activation */ std::nullopt,
+        /* compute_kernel_config */ ttml::core::ComputeKernelConfig::precise(),
+        /* core_grid */ ttnn::CoreGrid{7, 8},
+        /* output_tile */ std::nullopt);
+    ttml::core::MeshToXTensorVariant<float> composer = ttml::core::ConcatMeshToXTensor<float>(mesh_shape, 3);
+    auto xtensors_back = ttml::core::to_xtensor(mul_tensor, composer);
+    xt::xarray<float> mul_res = xt::linalg::dot(xtensor_a, xtensor_b);
```

Comment:
how dot works here? isn't b should be transposed for dot product to work?

Reply:
(128, 64) X (64, 256) => (128, 256)